File: event-streams/event-processor/src/main/java/com/scientiamobile/wurflmicroservice/eventprocessor/Processor.java
package com.scientiamobile.wurflmicroservice.eventprocessor;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.scientiamobile.wurfl.wmclient.Model;
import com.scientiamobile.wurfl.wmclient.WmClient;
import com.scientiamobile.wurfl.wmclient.WmException;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Processor class receives JSON events one by one from the standard input, uses the WURFL Microservice client
 * to enrich them with device detection data and prints them back to the standard output.
 */
public class Processor {

    public static void main(String[] args) throws IOException {

        List<String> errors = new ArrayList<>();

        String host = "localhost";
        switch (args.length) {
            case 0:
                // keeps default host
                break;
            case 2:
                host = args[1];
                break;
            default:
                System.out.print("Usage: event-processor-demo [--host <WURFL microservice serve IP>]");
                System.exit(1);
        }

        WmClient wmClient = null;
        try {
            wmClient = WmClient.create("http", host, "80", "");
            wmClient.setCacheSize(20000);
        } catch (WmException e) {
            System.out.println("Unable to connect to host: " + host);
            System.out.println(e.getMessage());
            System.exit(1);
        }
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        JsonReader reader = new JsonReader(new InputStreamReader(System.in));
        reader.setLenient(true);
        JsonToken token = reader.peek();
        while (token!= JsonToken.END_DOCUMENT && reader.hasNext()) {
            EnrichedEventData eventData = gson.fromJson(reader, EnrichedEventData.class);
            try {
                Model.JSONDeviceData device = wmClient.lookupUseragent(eventData.getUserAgent());
                eventData.setWurflCompleteName(device.capabilities.get("complete_device_name"));
                eventData.setWurflDeviceMake(device.capabilities.get("brand_name"));
                eventData.setWurflDeviceModel(device.capabilities.get("model_name"));
                eventData.setWurflFormFactor(device.capabilities.get("form_factor"));
                eventData.setWurflDeviceOS(device.capabilities.get("device_os") + " " + device.capabilities.get("device_os_version"));

                System.out.println(gson.toJson(eventData));
                token = reader.peek();

            } catch (WmException e) {
                errors.add("Cannot get device information for user-agent: " + eventData.getUserAgent());
            } catch (Exception e) {
                errors.add("Error: " + e.getMessage());
            }
            if (errors.size() > 0) {
                createErrorReport(errors);
            }
        }
        reader.close();
    }

    private static void createErrorReport(List<String> errors) {
        File reportFile = new File("error_reports.txt");
        if (!reportFile.exists()) {
            try {
                reportFile.createNewFile();
                OutputStream fw = new FileOutputStream(reportFile);
                BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fw, StandardCharsets.UTF_8));

                for (String s : errors) {
                    bw.write(s);
                    bw.newLine();
                }

                bw.flush();
                bw.close();

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: event-streams-dotnet/event-processor/Processor.cs
using Microsoft.AspNetCore.Http;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading;
using Wmclient;

namespace event_processor
{
    class Processor
    {
        static void Main(string[] args)
        {
            Console.InputEncoding = Encoding.UTF8;
            Console.OutputEncoding = Encoding.UTF8;
            var errors = new List<String>();

            String host = "localhost";
            switch (args.Length)
            {
                case 0:
                    // keeps default host
                    break;
                case 2:
                    host = args[1];
                    break;
                default:
                    Console.WriteLine("Usage: event-processor-demo [--host <WURFL microservice serve IP>]");
                    Thread.Sleep(2000);
                    Environment.Exit(1);
                    break;
            }

            WmClient wmClient = null;
            try
            {
                wmClient = WmClient.Create("http", host, "8080", "");
                wmClient.SetCacheSize(20000);
            }
            catch (WmException e)
            {
                Console.WriteLine("Unable to connect to host: " + host);
                Console.WriteLine(e.Message);
                Thread.Sleep(2000);
                Environment.Exit(1);
            }
           
            EnrichedEventData eventData = null;
            if (Console.IsInputRedirected)
            {
                using (StreamReader reader = new StreamReader(Console.OpenStandardInput(), Console.InputEncoding))
                {
                //using (var reader = new StreamReader(stream))
                //{
                    String json = reader.ReadToEnd();
                    while (json != null && json.Length > 0)
                    {
                        json = json.Trim();
                        var settings = new JsonSerializerSettings
                        {
                            MissingMemberHandling = MissingMemberHandling.Ignore,
                            NullValueHandling = NullValueHandling.Ignore,
                            StringEscapeHandling = StringEscapeHandling.EscapeNonAscii
                        };
                        eventData = JsonConvert.DeserializeObject<EnrichedEventData>(json, settings);
                        try
                        {
                            // Simulate data coming from an HTTP request
                            JSONDeviceData device = wmClient.LookupRequest(CreateRequest(eventData));
                            eventData.WurflCompleteName = device.Capabilities["complete_device_name"];
                            eventData.WurflDeviceMake = device.Capabilities["brand_name"];
                            eventData.WurflDeviceModel = device.Capabilities["model_name"];
                            eventData.WurflFormFactor = device.Capabilities["form_factor"];
                            eventData.WurflDeviceOS = device.Capabilities["device_os"] + " " + device.Capabilities["device_os_version"];

                            Console.WriteLine(JsonConvert.SerializeObject(eventData, Formatting.Indented, settings));
                            json = reader.ReadToEnd();
                        }
                        catch (WmException e)
                        {
                            Console.WriteLine(e.Message);
                            errors.Add("Cannot get device information for user-agent: " + eventData.UserAgent);
                        }
                        catch (Exception e)
                        {
                            Console.WriteLine(e.Message);
                            errors.Add("Error: " + e.Message);
                        }
                    }
                }
            }

            if (errors.Count > 0)
            {
                CreateErrorReport(errors);
            }
        }

        private static HttpRequest CreateRequest(EnrichedEventData eventData)
        {
            var request = new HttpRequestMock();
            request.Headers.Add("Accept", eventData.Accept);
            request.Headers.Add("User-Agent", eventData.UserAgent);
            request.Headers.Add("X-Ucbrowser-Device-Ua", eventData.UcBrowserDeviceUa);
            request.Headers.Add("X-Ucbrowser-Ua", eventData.UcBrowserUa);
            return request;
        }

        private static void CreateErrorReport(List<String> errors)
        {
            using (StreamWriter report = new StreamWriter(@"error_report.txt"))
            {
                foreach (string err in errors)
                {
                    report.WriteLine(err);
                }
            }
        }
    }
}

